web/modules/cm-lang-ftml/src/data/block.ts

```diff
@@ -0,0 +1,227 @@
+import {
+  EditorSvelteComponent,
+  type EditorSvelteComponentInstance
+} from "@wikijump/codemirror"
+import { type Completion } from "@wikijump/codemirror/cm"
+import { FTMLFragment } from "@wikijump/ftml-wasm-worker"
+import BlockTip from "../tips/BlockTip.svelte"
+import ModuleTip from "../tips/ModuleTip.svelte"
+import { aliasesFiltered, aliasesRaw } from "../util"
+import { Documentation } from "./data"
+import { htmlArgumentCompletions } from "./html-attributes"
+import { type Argument, type BlockConfiguration, type ModuleConfiguration } from "./types"
+
+const BlockTipFactory = new EditorSvelteComponent(BlockTip)
+const ModuleTipFactory = new EditorSvelteComponent(ModuleTip)
+
+export class ArgumentData {
+  declare name: string
+
+  declare type: `${"string" | "int" | "float" | "bool"}${"[]" | ""}`
```

Comment:
I'm amazed this works.

Reply:
Template string types are one of those things in TypeScript that shouldn't work but does.

In the `defaults.ts` file in the Fluent module, there is a type for Unit strings (like "millimeters" for "20mm") which is like 1800+ individual strings if you were to type it out as a union. That's because it supports doing `unit-per-unit`. There are some hilarious types like "feet-per-kelvin" because of that.